### ISSUE TITLE
How can we support authorization?

### ISSUE THREAD

**Issue by ieugen** [Distelli/graphql-apigen#14]
Hello,

I'm evaluating graphql-apigen for a production project - a SaaS applicanion. We need to filter data based on the current authenticated user. As per the Graphql spec recommends, this should be done in the business layer by passing a context object with the information needed to make the decision when loading data [1].

How does graphql-apigen support this use case? The generated code does not seem to offer a way to pass in the context. 

The only way I see how is to build the implementation objects with Request scope, but that seems quite out of the way. 

I believe we should be able to pass a context object to a `Resolver`. How did you solve this issue in you case?

Thanks, 

[1] http://graphql.org/learn/authorization/

**Comment by brimworks** [comment]
Hello,

This is a great question. I have not needed to solve this problem for my use-case yet.

My original thought was to use thread-locals to pass this context state, but that won't work if your using async support.

I like your idea of using Request scoped injected objects, but I do agree that going down that path is a bit out of the way.

How do you think apigen should handle this use-case?

Thanks,
-Brian

**Comment by bmsantos** [comment]
I had a similar need (Facebook like DataLoader's must follow a the query execution life cycle). In the end I've simply overloaded the [resolve method](https://github.com/bmsantos/graphql-apigen/blob/eb_graphql/apigen-deps/src/main/java/com/distelli/graphql/AsyncResolver.java). Hopefully there's a better solution for this.

**Comment by brimworks** [comment]
Thanks for sharing @bmsantos! BTW, I know @bmsantos has an open pull request for merging in async support, but I was waiting for the upstream graphql-java to release the async support. I did notice that graphql-java recently did a new release, does this mean I can merge in your (@bmsantos) changes?

Thanks,
-Brian

**Comment by bmsantos** [comment]
@brimworks, no. What happened was the last graphql-java release before moving to Java 8. The async branch requires Java 8 Futures and it should be available in the next release (3.0.0). We have to wait a wee bit longer. Sorry about that.

**Comment by ieugen** [comment]
@brimworks the solution with overloaded method seems good enough. 

@bmsantos how can we speed up the process? Could we have a 3.0.0-alfa release for the people who are a bit impatient (like me :D )? I would like to push some changes to the production environemnt based on graphql and can't do it without some form of authorization/filtering of data. Async support looks also promissing but we can live without it for a few months. 

@bmsantos , @brimworks : I can help whith this, just ping me some things that need solving and I'll give it some cycles.

**Comment by brimworks** [comment]
I don't want to rush into a change which involves passing an untyped Object into `resolve()`. It does make sense to pass in an extra context object to resolve(), but perhaps we can do it in a type-safe way?

**Comment by bmsantos** [comment]
@ieugen and @brimworks, I don't mind giving a shot at it but most probably will not happen until the next weekend.

**Comment by ieugen** [comment]
Great. I can wait until then. I have some days off. I've started working on a gradle plugin for apigen [1]. I think I can have something ready by then related to this. 

Regarding Context passing, we might add the class as a parameter when generating the classes. I'm thinking something like: 

````
 <plugin>
        <groupId>com.distelli.graphql</groupId>
        <artifactId>graphql-apigen</artifactId>
        <version>${apigen.version}</version>
        <configuration>
          <!-- This is only needed when using Guice -->
          <guiceModuleName>com.example.my.MyGuiceModule</guiceModuleName>
         <contextClass>com.example.context.MyContext</contextClass>
````

This should allow us to specify what we need in a type safe way. We don't even need to make sure the context is on the class path. It can come from some other library later on. 

This should be quite simple to implement. 

WDYT?

[1] https://github.com/netdava/graphql-apigen-gradle

**Comment by brimworks** [comment]
@ieugen yes, I was thinking something similar (allow specifying your own context). Perhaps the custom context class is "built" from the graphql.schema.DataFetchingEnvironment? In which case we could generate code that injects `com.example.context.MyContext.Factory` and calls `.create(graphql.schema.DataFetchingEnvironment)` which returns a `MyContext`?

Or, do you have other ideas on how the custom context is obtained?

Thanks,
-Brian

**Comment by ieugen** [comment]
Yeah. Im not that familiar with the internals yet, but I guess that sounds ok. We could generate an factory interface or have the user specify one. Not sure which is the right approach.

I'm in favor of generating the interface. I will know more when I start working with the code closer.

**Comment by bmsantos** [comment]
A side note: Executable schemas was just introduced in [graphql-java](https://github.com/graphql-java/graphql-java/pull/386/commits/de340f0805d56b2d4b1224ad84664fc3b247fcec).

**Comment by ieugen** [comment]
Thanks @bmsantos . That does change things a bit. Having that kind of capability inside graphql-java makes graphql-apigen less attractive since it will be hard to gather a community that can maintain it.

It might be a good thing if Apigen changes focus to do more things than generating classes. 

@brimworks What are your plans with graphql-apigen ? (discussion for a different thread maybe ?)

**Comment by brimworks** [comment]
The executable schema support sounds exciting. I'll have to dive into that more to see if it is a feasible alternative to graphql-apigen. However, at this current point in time, I intend to continue to support graphql-apigen regardless.

Thanks,
-Brian

**Comment by bmsantos** [comment]
@ieugen, sorry about it, but didn't have the time to get to it last weekend. The executable schema is a great addition to the spec but at this time, it still does not do the wiring of Resolvers/DataFetchers for you.

**Comment by ieugen** [comment]
@bmsantos : No worries. I looked a bit over executable schema and it is all runtime based. It does not generate any classes. I'm not sure I like that approach. 
I'm looking for something that can be compiled to be used externally, as a dependency. So graphql-apigen is still relevant.  It might be that graphql-apigen could reuse the exectuable schema from graphql-java directly and write the classes and wiring.

In the mean time I believe we can implement authorization via DataFetchers. 

DataFetchers receive a `DataFetchingEnvironment` that seems to be supplied on query execution. I can put the current user, roles and permission on the environment and the data fetcher will be able to perform the required operations. 

However, I don't see any way setting the data fetcher via graphql-apigen but I need to spend more time on this.

**Comment by bmsantos** [comment]
Started working on this.

@brimworks, @ieugen, why not just pass the actual **DataFetchingEnvironment**? There are other details in it that might be of interest to any given resolver.

Which approach do yo guys prefer at this stage?

**Comment by ieugen** [comment]
@bmsantos : Personally I don't like to introduce new concepts or API elements unless they are really neede. I did more work last night and it seems to me that `DataFetchingEnvironment` should be enough.

**Comment by brimworks** [comment]
I agree, just passing in the `DataFetchingEnvironment` is simpler. The next question is how exactly will this work? One idea is to have the generated interfaces extend an interface with a single method that takes as input the `DataFetchingEnvironment` and returns the same exact type. The MethodDataFetcher could then call that method. Something like this?

```java
public interface ResolveDataFetchingEnvironment<T> {
    public T resolve(DataFetchingEnvironment env);
}
```

The generated code would then be changed to extend this interface and implement a default that returns this:

```stringtemplate
public interface <model.name> extends ResolveDataFetchingEnvironment\<<model.name>> {
    public default <model.name> resolve(DataFetchingEnvironment env) {
        return this;
    }
    ...
}
```

...and `MethodDataFetcher` would now call this method:

```
    @Override
    public Object get(DataFetchingEnvironment env) {
        Object source = ( null != impl ) ? impl : env.getSource();
        if (source == null) return null;
        if (source implements ResolveDataFetchingEnvironment ) {
            source = ((ResolveDataFetchingEnvironment)source).resolve(env);
        }
        return getMethodViaGetter(source, env.getFieldType(), getFieldType(env.getParentType()), env.getArguments());
    }
```

Now to implement authorization, you could do something like this:

```java
import generated.MyObj;

public class MyObjImpl implements MyObj {
    private DataFetchingEnvironment env = null;
    @Inject private MyAuthorizer authorizer;
    // constructor used to wire this all together.
    @Inject protected MyObjImpl() {}
    // constructor used when we have the data fetching environment:
    private MyObjImpl(DataFetchingEnvironment env, MyObjImpl src) {
        // Copy over the fields of src:
        this.authorizer = src.authorizer;
        // Add in the env field:
        this.env = env;
    }
    @Override
    public MyObj resolve(DataFetchingEnvironment env) {
        return new MyObjImpl(env, this);
    }
    @Override
    public String getHello() {
        // Note that you could also AOP to implement this line of code:
        // https://github.com/google/guice/wiki/AOP
        authorizer.hasAccessToResource("hello", env);
        return "world";
    }
}
```

Does this sound good? Some of the favorable parts of this strategy are:

- Backwards compatible.
- Simple to implement (small change).
- Lots of flexibility.

Thanks,
-Brian

**Comment by ieugen** [comment]
It's a bit hard for me to follow all of this through since I'm not sure I have implemented the API corectly yet. 
What I can fallow seems to be ok with me. I like that it's a simple addition to the API. 

Do you have some code you could share? A simple example like hello schema or starwars schema would be great.

**Comment by bmsantos** [comment]
@brimworks, that's pretty much what I had in mind. I'll try to get something done this weekend. It shouldn't take long.

**Comment by bmsantos** [comment]
@brimworks, I've added an initial implementation following your guide lines. I must say that I didn't have much time to play with it but at least it is a start. @ieugen, maybe you can take it for a spin.

**Comment by bmsantos** [comment]
@brimworks, by the way, I had to make a couple of changes to have the tests to behave with maven 3.5.0. Which version of maven are you using to build this?

With 3.5.0 had to include:
```xml
    <dependency>
      <groupId>io.takari</groupId>
      <artifactId>commons-compress</artifactId>
      <version>1.12</version>
    </dependency>
```

and update @MavenVersions to 3.5.0:
```java
@RunWith(MavenJUnitTestRunner.class)
@MavenVersions({"3.5.0"})
public class ApiGenMojoTest {
```

**Comment by brimworks** [comment]
@bmsantos I use Maven 3.3.9.

I fixed some missing imports, added a unit test, and released this as version 2.3.0.

To add authorization to your own code you can follow this pattern (using Posts.createPost() as an example from this unit test):

[1] Override the resolve(DataFetchingEnvironment method) so that later method invocations can obtain a reference to it.

https://github.com/Distelli/graphql-apigen/blob/master/apigen/src/test/projects/posts/src/test/java/com/disteli/posts/PostsTest.java#L42

        @Override
        public MutatePostsImpl resolve(DataFetchingEnvironment env) {
            this.env = env;
            return this;
        }

[2] In your method that requires authorization, obtain the context from the DataFetchingEnvironment and perform authorization on it (here I use a simple string, but it could be a complex object):

https://github.com/Distelli/graphql-apigen/blob/master/apigen/src/test/projects/posts/src/test/java/com/disteli/posts/PostsTest.java#L48

```java
        @Override
        public Post createPost(MutatePosts.CreatePostArgs args) {
            if ( ! "authorized-user".equals(env.getContext()) ) {
                throw new java.security.AccessControlException("context MUST be authorized-user");
            }
            ...
        }
```

In your query code, pass the context object:

https://github.com/Distelli/graphql-apigen/blob/master/apigen/src/test/projects/posts/src/test/java/com/disteli/posts/PostsTest.java#L189

```java
        ExecutionResult result = graphQL.execute(
            "mutation{createPost(post:{title:\"NEW\" authorId:1}){title}}",
            "authorized-user");
```

Thanks,
-Brian
